src/resolver.rs

```diff
@@ -0,0 +1,490 @@
+                self.visit_statement(ctx, operator);
+                let operator_type_name = self
+                    .annotation_map
+                    .get_type(operator, self.index)
+                    .get_name();
+                if let Some(s) = parameters.as_ref() {
+                    let ctx = ctx.with_call(operator_type_name);
+                    self.visit_statement(&ctx, s);
+                }
+
+                if let Some(return_type) = self
+                    .index
+                    .find_return_type(operator_type_name)
+                    .and_then(|it| self.index.find_effective_type(it))
+                {
+                    self.annotation_map
+                        .annotate_type(statement, return_type.get_name());
+                }
+            }
+            _ => {
+                self.visit_statement_literals(ctx, statement);
+            }
+        }
+    }
+
+    /// annotate a literal statement
+    fn visit_statement_literals(&mut self, ctx: &VisitorContext, statement: &Statement) {
+        match statement {
+            Statement::LiteralBool { .. } => {
+                self.annotation_map.annotate_type(statement, BOOL_TYPE)
+            }
+            Statement::LiteralString { .. } => {
+                self.annotation_map.annotate_type(statement, STRING_TYPE);
+            }
+            Statement::LiteralInteger { value, .. } => {
+                self.annotation_map
+                    .annotate_type(statement, get_int_type_name_for(*value));
+            }
+            Statement::LiteralTime { .. } => {
+                self.annotation_map.annotate_type(statement, TIME_TYPE)
+            }
+            Statement::LiteralTimeOfDay { .. } => {
+                self.annotation_map
+                    .annotate_type(statement, TIME_OF_DAY_TYPE);
+            }
+            Statement::LiteralDate { .. } => {
+                self.annotation_map.annotate_type(statement, DATE_TYPE);
+            }
+            Statement::LiteralDateAndTime { .. } => {
+                self.annotation_map
+                    .annotate_type(statement, DATE_AND_TIME_TYPE);
+            }
+            Statement::LiteralReal { .. } => {
+                //TODO when do we need a LREAL literal?
+                self.annotation_map.annotate_type(statement, REAL_TYPE);
+            }
+            Statement::LiteralArray {
+                elements: Some(elements),
+                ..
+            } => {
+                self.visit_statement(ctx, elements.as_ref());
+                //TODO as of yet we have no way to derive a name that reflects a fixed size array
+            }
+            Statement::MultipliedStatement { element, .. } => {
+                self.visit_statement(ctx, element)
+                //TODO as of yet we have no way to derive a name that reflects a fixed size array
+            }
+            _ => {}
+        }
+    }
+}
+
+fn get_int_type_name_for(value: i64) -> &'static str {
```

Comment:
this implementation feels so wrong.
I dont think that we can guess the data-type of a literal just by looking at the value. The way we parse it, we will never see a negative number.
-2 becomes a UnaryExpression with a negate operator. So we will always end up with unsigned type-guessing?

Reply:
Should we handle this on the lexer and get a signed int parsed as a literal integer instead?

Reply:
ok for now everything is treated as DINT (32 bit signed int) or LINT if its bigger
i'll take a look at the negative terminals later ...

Reply:
Yes, we can open issues for both this and the array naming